packages/nx/src/plugins/js/lock-file/yarn-parser.ts

```diff
@@ -474,9 +474,10 @@ function mapSnapshots(
       const packageName = key.slice(0, key.indexOf('@', 1));
       let normalizedKey = key;
       if (isBerry && key.includes('@patch:') && key.includes('#')) {
+        const regEx = new RegExp(`@patch:${packageName}@(npm%3A)?(.*)$`);
         normalizedKey = key
           .slice(0, key.indexOf('#'))
-          .replace(`@patch:${packageName}@`, '@npm:');
+          .replace(regEx, '@npm:$2');
```

Comment:
The entries for yarn 3 follow this syntax:
`resolve@patch:resolve@1.22.8#~builtin<compat/resolve>`

The entries for yarn 4 follow this syntax:
`resolve@patch:resolve@npm%3A1.22.8#optional!builtin<compat/resolve>`